File: app/src/main/java/com/osrsd/cache/provider/HealthBarProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.Definition;
import com.osrsd.cache.def.HealthbarDefinition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;

@Slf4j
public final class HealthBarProvider {

    public static Definition decode(ByteBuffer buffer, HealthbarDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                buffer.getShort();
            } else if (opcode == 2) {
                definition.setField3277(buffer.get() & 0xff);
            } else if (opcode == 3) {
                definition.setField3278(buffer.get() & 0xff);
            } else if (opcode == 4) {
                definition.setField3283(0);
            } else if (opcode == 5) {
                definition.setField3275(buffer.getShort() & 0xffff);
            } else if (opcode == 6) {
                buffer.get();
            } else if (opcode == 7) {
                definition.setHealthBarFrontSpriteId(buffer.getShort() & 0xffff);
            } else if (opcode == 8) {
                definition.setHealthBarBackSpriteId(buffer.getShort() & 0xffff);
            } else if (opcode == 11) {
                definition.setField3283(buffer.getShort() & 0xffff);
            } else if (opcode == 14) {
                definition.setHealthScale(buffer.get() & 0xff);
            } else if (opcode == 15) {
                definition.setHealthBarPadding(buffer.get() & 0xff);
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/HealthBarProvider.kt
package com.osedit.cache.provider

import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.HealthBarDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object HealthBarProvider {

    private val log: Logger = LoggerFactory.getLogger(HealthBarProvider::class.java)

    fun decode(buffer: ByteBuffer, definition: HealthBarDefinition): Definition {
        do when (val opcode: Int = buffer.get().toInt() and 0xff) {
            1 -> definition.field3276 = buffer.short.toInt() and 0xffff
            2 -> definition.field3277 = buffer.get().toInt() and 0xff
            3 -> definition.field3278 = buffer.get().toInt() and 0xff
            4 -> definition.field3283 = 0
            5 -> definition.field3275 = buffer.short.toInt() and 0xffff
            6 -> definition.field3272 = buffer.get().toInt() and 0xff
            7 -> definition.healthBarFrontSpriteId = buffer.short.toInt() and 0xffff
            8 -> definition.healthBarBackSpriteId = buffer.short.toInt() and 0xffff
            11 -> definition.field3283 = buffer.short.toInt() and 0xffff
            14 -> definition.healthScale = buffer.get().toInt() and 0xff
            15 -> definition.healthBarPadding = buffer.get().toInt() and 0xff
            0 -> break
            else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
        } while (true)
        return definition
    }

}
